src/rust_connection/mod.rs

```diff
@@ -262,9 +248,86 @@ impl<R: ReadFD, W: WriteFD> RustConnection<R, W> {
             .send_request(ReplyFDKind::ReplyWithoutFDs)
             .expect("Sending a HasResponse request should not be blocked by syncs");
         inner.discard_reply(seqno, DiscardMode::DiscardReplyAndError);
-        write_all_vectored(write, &[IoSlice::new(&request)], Vec::new())?;
+        let inner = self.write_all_vectored(inner, write, &[IoSlice::new(&request)], Vec::new())?;
 
-        Ok(())
+        Ok(inner)
+    }
+
+    /// Write a set of buffers on a `writer`. May also read packets
+    /// from the server.
+    fn write_all_vectored<'a>(
+        &'a self,
+        mut inner: MutexGuardInner<'a>,
+        write: &mut W,
+        mut bufs: &[IoSlice<'_>],
+        mut fds: Vec<RawFdContainer>,
+    ) -> std::io::Result<MutexGuardInner<'a>> {
+        let mut partial_buf: &[u8] = &[];
+        while !partial_buf.is_empty() || !bufs.is_empty() || !fds.is_empty() {
+            let (can_read, can_write) = write.poll(true, true)?;
+            if can_write {
+                let mut count = if !partial_buf.is_empty() {
+                    write.write(partial_buf, &mut fds)?
+                } else {
+                    write.write_vectored(bufs, &mut fds)?
+                };
```

Comment:
Could `write` and `write_vectored` return `WouldBlock` here? (In theory: No, since we just polled for writability; in practice... dunno)

Reply:
If the OS is sane, it shoudn't

Reply:
Does that mean we have a problem on Mac OS? ;-)

Reply:
Is this happening on Mac OS?

Reply:
Nah, this was just meant as a generic "treating Mac OS as Unix leads to weird problems". I do not know about anything that behaves this way in `poll()`.